GVFS/GVFS.UnitTests/Tracing/QueuedPipeStringWriterTests.cs

```diff
@@ -0,0 +1,265 @@
+            Assert.AreEqual(QueuedPipeStringWriter.QueuedPipeStringWriterState.Stopped, eventCalls[1].State);
+        }
+
+        [TestCase]
+        public void GoodPipe_WritesDataAndRaisesStateHealthy()
+        {
+            const string inputMessage = "FooBar";
+            byte[] expectedData =
+            {
+                0x46, 0x6F, 0x6F, 0x42, 0x61, 0x72, (byte)'\n', // "FooBar\n"
+                0x46, 0x6F, 0x6F, 0x42, 0x61, 0x72, (byte)'\n', // "FooBar\n"
+                0x46, 0x6F, 0x6F, 0x42, 0x61, 0x72, (byte)'\n', // "FooBar\n"
+            };
+
+            string pipeName = Guid.NewGuid().ToString("N");
+
+            QueuedPipeStringWriter writer = new QueuedPipeStringWriter(() =>
+                new NamedPipeClientStream(".", pipeName, PipeDirection.Out));
+
+            // Capture event invocations
+            List<QueuedPipeStringWriter.QueuedPipeStringWriterStateEventArgs> eventCalls =
+                new List<QueuedPipeStringWriter.QueuedPipeStringWriterStateEventArgs>();
+            writer.StateChanged += (_, e) => eventCalls.Add(e);
+
+            using (TestPipeReaderWorker pipeWorker = new TestPipeReaderWorker(pipeName, PipeTransmissionMode.Byte))
+            {
+                // Start the pipe reader worker first and wait until the pipe server has been stood-up
+                // before starting the pipe writer/enqueuing messages because the writer does not wait
+                // for the pipe to be ready to accept (it returns and drops messages immediately).
+                pipeWorker.Start();
+                pipeWorker.WaitForReadyToAccept();
+                writer.Start();
+
+                // Try to write some dummy data
+                bool queueOk1 = writer.TryEnqueue(inputMessage);
+                bool queueOk2 = writer.TryEnqueue(inputMessage);
+                bool queueOk3 = writer.TryEnqueue(inputMessage);
+
+                // Wait until we've received all the sent messages before shuting down
+                // the pipe worker thread.
+                pipeWorker.WaitForRecievedBytes(count: expectedData.Length);
+                pipeWorker.Stop();
+                writer.Stop();
+
+                Assert.IsTrue(queueOk1);
+                Assert.IsTrue(queueOk2);
+                Assert.IsTrue(queueOk3);
+
+                byte[] actualData = pipeWorker.GetReceivedDataSnapshot();
+                CollectionAssert.AreEqual(expectedData, actualData);
+
+                // Should only receive one 'healthy' state change per successfully written message
+                Assert.AreEqual(2, eventCalls.Count);
+                Assert.AreEqual(QueuedPipeStringWriter.QueuedPipeStringWriterState.Healthy, eventCalls[0].State);
+                Assert.AreEqual(QueuedPipeStringWriter.QueuedPipeStringWriterState.Stopped, eventCalls[1].State);
+            }
+        }
+
+        private class TestPipeReaderWorker : IDisposable
```

Comment:
This seems like a lot of code to have in a unit test.  Can we instead abstract the `pipeClient` into an interface that is returned by `createPipeFunc` instead of the `NamedPipeClientStream` and use a mock of the interface?  Not sure how feasible this is or how much work it would be vs saved code here.  Let me know what you think.

Reply:
> This seems like a lot of code to have in a unit test.

Is the concern that this test is a bit large for a unit test? or that the test itself is not appropriate? If I understand correctly, it seems nice that we have a test that is exercising a bit more of the real components, even if it is a bit larger scope than desired for a unit test (i.e. would you prefer the test live "somewhere else" - not sure where that would be...).

Reply:
I guess my question would be what are we trying to test?  I looks like we are testing that the state change gets raised and the message gets "somewhere".  Right now the "somewhere" is a pipe and we have a bunch of code with a thread and locks and reset events which are all complex things to get right and understand but would it make it easy and less code if the "somewhere" was an interface so it could be a pipe class or a mock or something else? I'm not going to hold up the PR for this just something to think about because too many abstractions is also an issue that we don't want to get into.

Reply:
> I guess my question would be what are we trying to test? I looks like we are testing that the state change gets raised and the message gets "somewhere".

Correct.

> [..] if the "somewhere" was an interface so it could be a pipe class or a mock or something else? I'm not going to hold up the PR for this just something to think about because **too many abstractions** is also an issue that we don't want to get into.

So in this case I'd need to wrap `NamedPipeClientStream` which includes construction, connection, and sending/writing functions. This feels a little bit like too many abstractions.
Ideally I could use a `Stream` (and then pass in a `MemoryStream` or something in testing.. but only the pipe streams have the notion of connection).

The `TestPipeReaderWorker` is mocking out the telemetry service behaviour which is to be running ahead of mount, waiting for a connection and draining the pipe until it gets shutdown.

Reply:
Okay sounds good. We might want to eventually move this outside of this class so that other tests that deal with pipes can use it.

Reply:
Absolutely. At the moment the test class has only been designed with these use-cases in mind so I'd rather not pull it up-and-out yet until it's needed. Then we can worry about if it's general purpose. 😁